### PR TITLE
Major clean up and subscription -> pokes refactor

### PR DESCRIPTION
**Pull request by ashelkovnykov**
**Status**

Backend and frontend changes made. Tested by @ashelkovnykov and @bonbud-macryg .

**Resolves**

#69 
#70 
#78 
#80 
#104 
#115

### CODE REVIEW

**Comment by bonbud-macryg** [comment]
Frontend work that needs to be done, just looking at these issues before review:
* #69 #70
  * Update FE interface with pokes and updates.
* #115
  * Update FE's "Undo" button to reflect when you can and can't undo.

File: src/backend/chess/lib/chess.hoon [hunk 1]
```diff
@@ -2,14 +2,42 @@
...
+::  but useful for testing.
+++  empty-board
+  ^-  chess-board
+  ~
+::
+::  a test game
+::  a game that ends in four moves with black side winning
+++  fools-mate
+  ^-  chess-game
+  :*  game-id=~2021.2.2..11.01.55
```
Comment (ashelkovnykov) [inline src/backend/chess/lib/chess.hoon]: Cut off part of the game ID, need to put it back

File: src/backend/chess/lib/chess.hoon [hunk 2]
```diff
@@ -1330,125 +1418,4 @@
...
-    ::
-        %kingside
-      ?-  player
-        %white  ['e1' 'g1']
-        %black  ['e8' 'g8']
-      ==
-    ==
-  ==
---
+--
```
Comment (ashelkovnykov) [inline src/backend/chess/lib/chess.hoon]: Need trailing newline

File: src/backend/chess/app/chess.hoon [hunk 3]
```diff
@@ -15,17 +18,18 @@
...
 +$  state-1
   $:  %1
-      games=(map @dau active-game-state)
-      archive=(map @dau chess-game)
+      games=(map game-id active-game-state)
+      potential-states=(map game-id (list (pair active-game-state card)))
+      archive=(map game-id chess-game)
       challenges-sent=(map ship chess-challenge)
       challenges-received=(map ship chess-challenge)
       rng-state=(map ship chess-commitment)
```
Comment (bonbud-macryg) [inline src/backend/chess/app/chess.hoon:35]: What's the thinking behind `potential-states` and `future-states`?

**Review by bonbud-macryg — COMMENTED** [review]
Big improvement thus far.

File: src/backend/chess/app/chess.hoon [hunk 4]
```diff
@@ -76,570 +61,946 @@
...
+          :_  this
+          ::  attempt to accept game
+          ::  handle our end on ack
+          :~  :*  %pass
+                  /poke/game/(scot %da game-id)/init
+                  %agent
+                  [who.action %chess]
+                  %poke
+                  %chess-action
+                  !>([%game-accepted game-id our-side])
```
Comment (bonbud-macryg) [inline src/backend/chess/app/chess.hoon]: Looks like there are some marks still to update throughout? I think `%chess-action` here should be `%chess-agent-action` and `%game-accepted` should be `%challenge-accepted`.

**Comment by bonbud-macryg** [comment]
As of `bdfd9dd`:

* Practice board seems fully functional.
* I'm seeing that the scry to %pals loudly breaks if %pals isn't installed on the fakeship, but this is a separate issue.
* Trying to send a challenge to another player breaks with the following. (EDIT 1 : Could be my fakeship being out of date; looking into it.)

```
clay: read-at-aeon fail [desk=%chess care=%c case=[%da p=~2023.8.18..21.24.04..1d19] path=/json/chess-user-action]
-find.^json
/mar/json/hoon::[11 9].[11 14]>
/mar/json/hoon::[11 5].[11 14]>
/mar/json/hoon::[11 1].[26 3]>
/mar/json/hoon::[10 1].[26 3]>
/mar/json/hoon::[9 1].[26 3]>
/mar/json/hoon::[8 1].[26 3]>
[%error-building /mar/json/hoon]
[%error-building-cast %json %chess-user-action]
[%error-building-tube %json %chess-user-action]
```

EDIT 2: Failure to send the error was the result of the outdated fakeship, but the other ship fails to receive.

**Comment by bonbud-macryg** [comment]
We can send challenges but opponents don't receive them. I can tell this is out of date:

`/app/chess` lines 81-89
```
::  send new challenge
:~  :*  %pass
        /poke/challenge/send
        %agent
        [who.action %chess]
        %poke
        %chess-challenge
        !>(challenge.action)
==  ==
```

But having taken a couple of stabs at it I'm not sure exactly what needs to happen to update the poke and others like it. I know roughly what the second of these lines in `/mar/user-action` is doing, not sure about the first. Of course whole file is new, but I think the idea of these two lines is new to me.

```
%-  chess-user-action:chess
=/  head  ((ot:dejs ~[[%chess-user-action so:dejs]]) jon)
```

I think it's been so long since I've looked at the frontend I need to refresh myself on what the data actually looks like as it moves from FE and through the Urbit side, which I can do by printing stuff out at different stages.

File: src/backend/chess/app/chess.hoon (hunk 3 above)
Comment (ashelkovnykov) [reply src/backend/chess/app/chess.hoon:35]: If we enable pre-moves (which the Chessground frontend supports), we can allow users to queue up several moves. Advanced players often run into situations where they know exactly what moves the opponent will make and want to queue up moves in advance so as not to waste time off the clock.

**Comment by ashelkovnykov** [comment]
I've update the branch for my latest changes, and updated the PR description.

@bonbud-macryg I tried to preserve your work as independent commits at first, but it became intractable with so much needing to be done, so I just ended up squashing everything since the last change to the backend which makes sense as an independent commit into one giant commit.

**Comment by bonbud-macryg** [comment]
As of `b651eb2`:

* Sending a challenge works
* **Can't decline challenge**
* Accepting a challenge works
* Moving pieces works, taking pieces works
* Game panel works: can browse previous states, view resets on new move from opponent
* Sending draw offer works
* **Declining draw offer doesn't fully update frontend or opponent state**
* Sending undo request works
* **Declining undo request doesn't fully update frontend or opponent state**
* Check works
* Checkmate works on winner's end
* **Checkmate doesn't make the game vanish on the loser's end**
* Resign works
* Practice board works
* Custom visuals work

**Untested**
* Pals integration

**Comment by bonbud-macryg** [comment]
TODOs, with more accurate descriptions of the bugs.

- [x] Can't decline challenge
  * Fixed on the backend, ~~untested on frontend~~ works on frontend, but doesn't update instantly

- [x] Can't decline draw
  * Could be a frontend issue; `/mar/chess/update` looks fine.

```
"conversion failed from chess-update to json"
/sys/vane/eyre/hoon:<[1.874 9].[1.874 72]>
/sys/vane/eyre/hoon:<[1.874 29].[1.874 71]>
/sys/vane/eyre/hoon:<[1.874 30].[1.874 71]>
/sys/vane/eyre/hoon:<[1.874 43].[1.874 70]>
/mar/chess/update/hoon:<[17 5].[144 7]>
/mar/chess/update/hoon:<[85 9].[88 11]>
/mar/chess/update/hoon:<[86 9].[88 11]>
/mar/chess/update/hoon:<[87 13].[87 51]>
/mar/chess/update/hoon:<[87 23].[87 50]>
/mar/chess/update/hoon:<[87 27].[87 49]>
```

- [x] Can't accept or decline undo
  * Both accepting and declining undo requests get the same error as the `%draw-declined` error, at different places in `/mar/chess/update`
  
```
"conversion failed from chess-update to json"
... (in %accepted-undo and %declined-undo respectively)
```
  
- [x] Checkmate issue
  * Not sure how this interacts with #2

**Comment by ashelkovnykov** [comment]
Updated PR description for latest status

**Comment by bonbud-macryg** [comment]
Added a Revoke Draw button and looks good on the frontend, but actually clicking Revoke Draw does nothing because it gets a `"conversion failed from chess-update to json"` error in the Dojo.

(Edit: fixed on next commit.)

**Comment by bonbud-macryg** [comment]
Tested pals integration, seems to work as intended.

**Comment by bonbud-macryg** [comment]
Crashing on last move:
* Has nothing to do with which sides the players are
* Winner receives the result in `+on-agent`
* Loser never receives the result on backend

**Pull request merged**